nixos/modules/services/cluster/druid/default.nix

```diff
@@ -0,0 +1,279 @@
+        "druid.service" = "servicename";
+      };
+    };
+
+    jdk = mkPackageOption pkgs "JDK" {
+      default = [ "jdk17_headless" ];
+    };
+
+    jvmArgs = mkOption {
+      type = types.str;
+      default = "";
+      description = "Arguments to pass to the JVM";
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Open firewall ports for ${serviceName}.";
+    };
+
+    internalConfig = mkOption {
+      default = { };
+      type = types.attrsOf types.anything;
+      internal = true;
+      description = "Internal Option to add to runtime.properties for ${serviceName}.";
+    };
+  };
+
+
+
+  druidServiceConfig =
+    { name
+    , serviceOptions ? cfg."${name}"
+    , allowedTCPPorts ? [ ]
+    , tmpDirs ? [ ]
+    , extraConfig ? { }
+    }: (
+      mkIf serviceOptions.enable (mkMerge [
+        {
+          systemd = {
+            services."druid-${name}" = {
+              after = [ "network.target" ];
+
+              description = "Druid ${name}";
+
+              environment = {
+                DRUID_JAVA_HOME = "${serviceOptions.jdk}";
+              };
+
+              wantedBy = [ "multi-user.target" ];
+
+              inherit (serviceOptions) restartIfChanged;
+
+              path = [ "${serviceOptions.jdk}" ];
+
+              script =
+                let
+                  cfgFile = fileName: properties: pkgs.writeTextDir fileName
+                    (concatStringsSep "\n" (mapAttrsToList (n: v: "${n}=${toString v}") properties));
+
+                  commonConfigFile = cfgFile "common.runtime.properties" cfg.commonConfig;
+
+                  configFile = cfgFile "runtime.properties" (serviceOptions.config // serviceOptions.internalConfig);
+
+                  extraClassPath = concatStrings (map (path: ":" + path) cfg.extraClassPaths);
+
+                  extraConfDir = concatStrings (map (dir: ":" + dir + "/\*") cfg.extraConfDirs);
+                in
+                ''
+                  ${cfg.package}/bin/run-java -Dlog4j.configurationFile=file:${cfg.log4j} \
+                    -Ddruid.extensions.directory=${cfg.package}/extensions \
+                    -Ddruid.extensions.hadoopDependenciesDir=${cfg.package}/hadoop-dependencies \
+                    -classpath  ${commonConfigFile}:${configFile}:${cfg.package}/lib/\*${extraClassPath}${extraConfDir} \
+                    ${serviceOptions.jvmArgs} \
+                    org.apache.druid.cli.Main server ${name}
+                '';
+
+              serviceConfig = {
+                User = "druid";
+                SyslogIdentifier = "druid-${name}";
+                Restart = "always";
+              };
+            };
+
+            tmpfiles.rules = concatMap (x: [ "d ${x} 0755 druid druid" ]) (cfg.commonTmpDirs ++ tmpDirs);
+          };
+          networking.firewall.allowedTCPPorts = mkIf
+            ((attrByPath [ "openFirewall" ] false serviceOptions))
```

Comment:
Maybe a stupid question..But should this not be `true`?

Reply:
We are using attrByPath to set the default value of openFirewall to false if the property was not set expliclity in serviceOptions.

Reply:
ah makes sense, thanks